tests/installation/validation/iscsi_ibft.pm

```diff
@@ -0,0 +1,171 @@
+# SUSE's openQA tests
+#
+# Copyright © 2016-2019 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Test module to verify iBFT configuration
+# iSCSI Boot Firmware Table (iBFT) is a mechanism for the iSCSI tools to
+# extract from the machine NICs the iSCSI connection information so that they
+# can automagically mount the iSCSI share/target.  Currently the iSCSI
+# information is hard-coded in the initrd. The /sysfs entries are read-only
+# one-name-and-value fields.
+# Maintainer: Martin Loviska <mloviska@suse.com>
+
+use base "opensusebasetest";
+use strict;
+use warnings;
+use testapi;
+use Scalar::Util 'looks_like_number';
+use Data::Dumper;
+
+my $ibft_expected = {
+    ethernet => {
+        flags       => 3,
+        prefix_len  => 24,
+        subnet_mask => '255.255.255.0',
+        gateway     => '10.0.2.2',
+        ip_addr     => '10.0.2.15',
+        mac         => '',
+    },
+    initiator => {
+        initiator_name => 'iqn.2010-04.org.ipxe:00000000-0000-0000-0000-000000000000',
+        flags          => 3,
+    },
+    target => {
+        target_name => '',
+        flags       => 3,
+        port        => '3260',
+        ip_addr     => '',
+        chap_type   => 0
+    },
+    acpi_header => {
+        oem_table_id => 'iPXE',
+        oem_id       => 'FENSYS',
+        signature    => 'iBFT'
+    },
+    backstore => {
+        model  => q/'FILEIO|IBLOCK'/,
+        vendor => q/'LIO-ORG'/
+    }
+};
+
+$ibft_expected->{target}->{target_name} = get_required_var('NBF');
+$ibft_expected->{target}->{ip_addr}     = get_required_var('WORKER_HOSTNAME');
+$ibft_expected->{ethernet}->{mac}       = get_required_var('NICMAC');
+
+# bit 0: block valid flag should be set
+# bit 1: Firmware boot selected flag should be set
+sub is_ibft_boot_flag {
+    my $received = shift;
+    my $mask     = oct 0b0000_0011;
+    diag("Comparing flag bits...\nGot: $received\nExpected: $mask");
+    if (($received & $mask) == 3) {
+        return !!1;
+    }
+    return !!0;
+}
+
+sub compare {
+    my ($got, $expected, $msg) = @_;
+    diag($msg) if defined($msg);
+    if (looks_like_number($expected) && looks_like_number($got)) {
+        return !!($expected == $got);
+    } else { return !!($expected eq $got); }
```

Comment:
Maybe you could use `Assert::Equals`, like in https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/validate_lvm_encrypt.pm#L72 ? We got this package installed on the workers by now.

Reply:
ah, I have forgotten about it. But I like my compare, it is even sending the data to os-autoinst.log.
Lovely!
http://eris.suse.cz/tests/12585/file/autoinst-log.txt

```
[2019-03-20T15:06:57.032 CET] [debug] 
Comparing: acpi_header->oem_table_id
Got: iPXE
Expected: iPXE
```

Reply:
Assert equals will do same in case of mismatch and you will be able to provide message explaining the comparison. But that's up to you. The only thing that you don't need to convert comparisons to the boolean context with `!!` trick, as it's already boolean context.

Reply:
It is better to replace my compare by Assert::Equals. As long as it is in cpan, I assume Assert::Equals is properly tested. :)

Reply:
It's really nice wrapper and in case of failures it's well visualized in openQA, so it worked better than I have expected.

Reply:
I have not simulated any failure yet. I will trigger one run which should fail at this point

Reply:
Frankly said, I like more this output:
http://eris.suse.cz/tests/12548#step/iscsi_ibft/67

Mainly it will check all parameters, despite a mismatch.

Reply:
I mean we will check them anyway, but most of the time we don't care about certain check, unless it fails. And in case of failure both ways provide same information, receive value, expected and the message which explains it.